code/scripts/DirectoryController.hs

```diff
@@ -0,0 +1,112 @@
+iterator folder = do
+  setCurrentDirectory (folderPath folder)
+  rawContents <- listDirectory (folderName folder)
+
+  let bakedContents = (folders,files)
+      folders = map (createFolder workingDirectory currentDrasilPack) folderNames
+      files = map (createFile workingDirectory currentDrasilPack) fileNames
+
+      folderNames = sort $ rawContents \\ filter (isInfixOf ".") rawContents
+      fileNames = sort $ filter (isSuffixOf ".hs") rawContents
+
+      workingDirectory = getFolderPath folder
+      currentDrasilPack = folderDrasilPack folder
+  return bakedContents
+
+-- recursively searches for all folders + files in a directory using iterator
+finder :: Folder -> IO [File]
+finder folder = do
+  rawData <- iterator folder
+
+  let rawFolders = fst rawData
+  folders <- verifyDirectories rawFolders
+  rawFiles <- mapM finder folders
+
+  let bakedFiles
+        | null rawFiles = []
+        | otherwise = concat rawFiles
+
+  let files 
+        | null folders = snd rawData
+        | otherwise = bakedFiles ++ snd rawData
+  return files
+
+-- gets all drasil- packages + filepaths in a list of folder data types
+getDirectories :: FilePath -> FilterPrefix -> IO [Folder]
+getDirectories directoryPath filterPrefix = do
+  -- all raw directory contents
+  all <- listDirectory directoryPath
+  -- raw drasil- package directories + package names
+  let rawPackages = sort $ filter (isPrefixOf filterPrefix) all
+      packageNames = map (\\"drasil-") rawPackages
+  -- convert list of directories into folder data types
+      directories = zipWith (createFolder directoryPath) packageNames rawPackages
+  return directories
+
+-- verifies that each folder/directory exists
+verifyDirectories :: [Folder] -> IO [Folder]
+verifyDirectories rawFolders = do
+  let rawDirectories = map getFolderPath rawFolders
+  boolFolders <- mapM doesDirectoryExist rawDirectories
+  let verifiedDirectories = snd $ partition nullFolder (zipWith fBool boolFolders rawFolders)
+  return verifiedDirectories
+
+-- combines lists of Booleans and Folders (if True, Folder exists)
+fBool :: Bool -> Folder -> Folder
+fBool b f = if b then f else createFolder "" "" ""
+
+-- checks if a folder is null (empty)
+nullFolder :: Folder -> Bool
+nullFolder folder = empty where
+  empty = all null [folderName folder,folderPath folder,folderDrasilPack folder]
+
+-- combines folder name with filepath (for testing if directory exists)
+joins :: [FilePath] -> FilePath
+joins (a:b:_) = b ++ "/" ++ a
+
+-- creates new folder path with folder name + path (to extract folder contents)
+getFolderPath :: Folder -> FilePath
+getFolderPath folder = folderPath folder ++ "/" ++ folderName folder
+
+-- strips pathSuffix from filepath, returning stripped filepath 
+-- strip occurs iff pathSuffix is suffix of filepath
+stripPath :: FilePath -> FilePath -> IO FilePath
+stripPath filePath pathSuffix = do
+  let strippedPath
+        | pathSuffix `isSuffixOf` filePath = take (lfP - lpS) filePath
```

Comment:
The take portion could likely be replaced with [`takeDirectory`](https://hackage.haskell.org/package/filepath-1.4.2.1/docs/System-FilePath-Posix.html#v:takeDirectory)

Reply:
I might just leave the take part as it is, and simple use `takeDirectory` instead of `stripPath` as is ideal, if that's okay. Or is it better for the take part to be replaced with `takeDirectory` as is ideal, and just use `stripPath` all throughout these scripts?

Reply:
Or should I just remove `stripPath` altogether? I've replaced its only application with `takeDirectory`, and I don't see any useful applications for `stripPath` anywhere else in these scripts in the near future.

Reply:
You should use standard Haskell library functions as much as possible, especially when dealing with operating-system related things, as they are much more likely to have been thoroughly debugged and work cross-platform.

Reply:
Alright then. In that case, should I remove `stripPath` altogether, or should I just comment it all out?

Reply:
Right!